### DISCUSSION TITLE
How to enable undo/redo operations on the Tree Editor

### DISCUSSION THREAD

**Discussion by agonzalezgaliana** [eclipse-emfcloud/emfcloud#156] [Theia Tree Editor] [answered]
I'm currently doing good progress on using the Tree Editor with my Model Server. So far I've followed the Coffee Editor as template for my implementation. I'm able to show my model, create and delete different kind of nodes and save my changes on the Model Server.

The save operation was easy to implement since my widget implements `NavigatableTreeEditorWidget` and hence the `Saveable` interface.

I would expect to have a similar mechanism in my widget to implement the logic for undo/redo, However I could not find any reference to undo or redo on the `@eclipse-emfcloud/theia-tree-editor` module.

Is the undo/redo operations available for the Tree Editor? if so, how to use them?

Many thanks in advance.

**Comment by ndoschek** [comment, marked as answer]
Hi @agonzalezgaliana!
Great to hear that you use the tree editor! :deciduous_tree: 

Basically, undo/redo is available via the common Theia undo/redo commands from the menu `Edit > Undo / Redo`, but the tree-editor widget does not handle those commands yet.
For example, the (GLSP) Diagram widget overrides the global Theia undo/redo commands and dispatches a dedicated `Undo/RedoAction` which is handled by the server.
This could also be picked up in the same manner for the tree-editor, and instead of dispatching an Action, directly call the `modelServerClient` with `undo`/`redo`.
See for example in `sprotty-theia`:
- https://github.com/eclipse/sprotty-theia/blob/bc7b3d783414bebf67a45905d756f286d181f34e/src/theia/diagram-commands.ts#L200
- https://github.com/eclipse/sprotty-theia/blob/bc7b3d783414bebf67a45905d756f286d181f34e/src/theia/diagram-keybinding.ts#L40

Hope that helps and best wishes,
Nina

PS: Please feel free to open an issue and/or in case you would like to see it in the theia-tree-editor base framework, thanks!

**Reply by agonzalezgaliana** [reply]
Hi @ndoschek thanks a lot for your helpful response :grinning:. I was able to enable the undo/redo operations by following the same implementation from Sprotty.
